lib/pbench/cli/server/shell.py

```diff
@@ -40,42 +42,135 @@ def find_the_unicorn(logger: Logger):
         )
 
 
-def generate_crontab_if_necessary(
-    crontab_dir: str, bin_dir: Path, cwd: str, logger: Logger
-) -> int:
-    """Generate and install the crontab for the Pbench Server.
+class PbenchIndexer:
+    """Managed the context of the indexer subprocesses."""
 
-    If a crontab file already exists, no action is taken, otherwise a crontab
-    file is created using `pbench-create-crontab` and then installed using the
-    `crontab` command.
+    def __init__(self, bin_dir: Path, cwd: Path, logger: Logger):
+        """Construct the subprocess tracker state.
 
-    If either of those operations fail, the crontab file is removed.
+        Args:
+            bin_dir: Where to find the `pbench-index` command
+            cwd: Where to set the command working directory
+            logger: Python logger for feedback.
+        """
+        self.bin_dir = bin_dir
+        self.cwd = cwd
+        self.logger = logger
+        self.indexer: Optional[subprocess.Popen] = None
+        self.reindexer: Optional[subprocess.Popen] = None
+        self.shutoff: Event = Event()
+        self.reaper: Optional[Thread] = None
 
-    Return 0 on success, 1 on failure.
-    """
-    ret_val = 0
-    crontab_f = Path(crontab_dir) / "crontab"
-    if not crontab_f.exists():
-        os.environ["PATH"] = ":".join([str(bin_dir), os.environ["PATH"]])
-        # Create the crontab file from the server configuration.
-        cp = subprocess.run(["pbench-create-crontab", crontab_dir], cwd=cwd)
-        if cp.returncode != 0:
-            logger.error(
-                "Failed to create crontab file from configuration: {}", cp.returncode
-            )
-            ret_val = 1
-        else:
-            # Install the created crontab file.
-            cp = subprocess.run(["crontab", f"{crontab_dir}/crontab"], cwd=cwd)
-            if cp.returncode != 0:
-                logger.error("Failed to install crontab file: {}", cp.returncode)
-                ret_val = 1
-    if ret_val != 0:
-        crontab_f.unlink(missing_ok=True)
-    return ret_val
-
-
-def run_gunicorn(server_config: PbenchServerConfig, logger: Logger) -> int:
+    def start_indexer(self):
+        """Isolate starting the primary indexer as a subprocess.
+
+        NOTE: we specify the 'python3' command explicitly rather than relying
+        on the shebang to avoid the intermediate `env python` command which
+        complicates reaping.
+        """
+        self.indexer = subprocess.Popen(
+            ["python3", self.bin_dir / "pbench-index"], cwd=self.cwd
+        )
```

Comment:
I thought of that, and decided it probably wasn't worthwhile as they log their own errors.

Reply:
That's fine.

But, as a general matter of principal, since the subprocess by default inherits `stdin`, `stdout`, and `stderr`, it would be good to be explicit about how we dispose of them.  I recommend setting `stdin` to `DEVNULL`, because we really don't want the subprocess "stealing" input from the parent.  Likewise, unless we're sure we like where the parent's `stdout` and `stderr` are going, I suggest redirecting the outputs of the subprocess to a known location.